cpp/tests/collect_list/collect_list_test.cpp

```diff
@@ -0,0 +1,594 @@
+/*
+ * Copyright (c) 2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf_test/base_fixture.hpp>
+#include <cudf_test/column_utilities.hpp>
+#include <cudf_test/column_wrapper.hpp>
+#include <cudf_test/cudf_gtest.hpp>
+#include <cudf_test/type_lists.hpp>
+
+#include <cudf/aggregation.hpp>
+#include <cudf/detail/aggregation/aggregation.hpp>
+#include <cudf/rolling.hpp>
+#include <cudf/table/table_view.hpp>
+#include <cudf/utilities/bit.hpp>
+#include <src/rolling/rolling_detail.hpp>
+
+#include <thrust/iterator/constant_iterator.h>
+
+#include <algorithm>
+#include <vector>
+
+struct CollectListTest : public cudf::test::BaseFixture {
+};
+
+template <typename T>
+struct TypedCollectListTest : public CollectListTest {
+};
+
+using TypesForTest = cudf::test::
+  Concat<cudf::test::IntegralTypes, cudf::test::FloatingPointTypes, cudf::test::DurationTypes>;
+
+TYPED_TEST_CASE(TypedCollectListTest, TypesForTest);
+
+TYPED_TEST(TypedCollectListTest, BasicRollingWindow)
+{
+  using namespace cudf;
+  using namespace cudf::test;
+
+  using T = TypeParam;
+
+  auto const input_column = fixed_width_column_wrapper<T, int32_t>{10, 11, 12, 13, 14};
+
+  auto const prev_column = fixed_width_column_wrapper<size_type>{1, 2, 2, 2, 2};
```

Comment:
Is there a possibility that `prev_column` and `foll_column` to have negative values ?

Reply:
Not supporting that at the moment, in any of the rolling-window functions. This will require a fair bit of testing.
It's on the cards for update later. There are Spark use-cases where the user might have preceding/following identifying a window situated completely "before" the current row. At that point, the fact that `preceding` includes the current row obfuscates the whole calculation. :/

I'd like for those cases to be tackled separately.